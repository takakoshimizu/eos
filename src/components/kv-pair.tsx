import React, { FunctionComponent } from 'react';
import styled from 'styled-components';

const $KvPair = styled.div.attrs({
	tabIndex: 0
})`
	display: flex;
	font-size: 1.2rem;
`;

const $Key = styled.div`
	font-weight: 600;
	margin-right: 1rem;
	flex: 0 0 7em;
`;

const $Value = styled.div`
	flex: 1;
`;

type Props = {
	keyName: string;
	children: string;
}

export const KvPair: FunctionComponent<Props> = ({ keyName, children }) => (
	<$KvPair>
		<$Key>{keyName}</$Key>
		<$Value>{children}</$Value>
	</$KvPair>
)
