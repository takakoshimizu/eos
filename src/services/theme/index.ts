export const theme: Theme = {
	light: {
		neutral: '#fefefe',
		text: '#111111',
		accent: 'blue'
	},
	dark: {
		neutral: '#111111',
		text: '#fefefe',
		accent: 'darkblue'
	},
	font: {
		primary: 'Helvetica, sans-serif'
	}
};

type Shade = 'light' | 'dark';
export type Theme = { [key: string]: Record<string, string> } 
type PropsWithTheme = { theme: Theme };

export const color = (shade: Shade, color: string) => (props: PropsWithTheme): string => props.theme[shade][color];
