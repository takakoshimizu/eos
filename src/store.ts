import { combineReducers, applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import promiseMiddleware from 'redux-promise';

import blocks from './services/blocks/reducer';

const reducers = combineReducers({
	blocks
});

const composeEnhancers = composeWithDevTools;
export const store = createStore(reducers, composeWithDevTools(applyMiddleware(promiseMiddleware)));
