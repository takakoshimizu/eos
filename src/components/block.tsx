import React, { useState, FunctionComponent } from 'react';
import styled from 'styled-components';

import { KvPair } from './kv-pair';

import { color, shadow } from '../services/theme';

const $Block = styled.li`
	padding: 1rem;
	margin: 1rem 0;
	border-radius: 6px;
	list-style: none;
	cursor: default;

	box-shadow: ${shadow('z1')};
	transition: box-shadow 0.3s ease-in-out;

	background-color: ${color('light', 'card')};
	border-top: 4px solid ${color('light', 'accent')};
	@media (prefers-color-scheme: dark) {
		background-color: ${color('dark', 'card')};
		border-top-color: ${color('dark', 'accent')};
	}

	&:hover {
		box-shadow: ${shadow('z2')};
	}

	button {
		margin-top: 1em;
	}
`;

const $Raw = styled.pre`
	max-width: 100%;
	max-height: 50vh;
	overflow: auto;
`;

type Props = {
	block: Block;
}

export const Block: FunctionComponent<Props> = ({ block }) => {
	const [isOpen, setOpen] = useState(false);
	
	return (
		<$Block>
			<KvPair keyName="ID">{block.id}</KvPair>
			<KvPair keyName="Timestamp">{block.timestamp}</KvPair>
			<KvPair keyName="Actions">{block.transactions.length.toString()}</KvPair>

			<button onClick={() => setOpen(!isOpen)}>{'Toggle details'}</button>

			{isOpen && (
				<$Raw>{JSON.stringify(block, null, 2)}</$Raw>
			)}
		</$Block>
	);
};
