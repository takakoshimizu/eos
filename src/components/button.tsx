import styled from 'styled-components';
import { color, shadow } from '../services/theme';
import shader from 'shader';

export const Button = styled.button`
	padding: .5em 1em;
	border-radius: 6px;
	border: 0;

	box-shadow: ${shadow('z1')};

	background-color: ${color('accent')};
	color: ${color('buttonText')};

	transition: 
		background-color .3s ease-in-out,
		box-shadow .3s ease-in-out,
		transform .2s linear;

	&:hover {
		box-shadow: ${shadow('z2')};
		background-color: ${p => shader(color('accent')(p), 0.2)};
		transform: translateY(-2px);
	}
`;
