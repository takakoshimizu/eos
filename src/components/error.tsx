import React, { FunctionComponent } from 'react';
import styled from 'styled-components';
import { color, shadow } from '../services/theme';

const $Error = styled.div`
	margin: 1rem 0;
	padding: 1rem;
	text-align: center;
	font-size: 1.2em;

	border-radius: 6px;
	background-color: ${color('card')};
	box-shadow: ${shadow('z1')};
	color: ${color('error')};
`;

type Props = {
	isVisible: boolean;
	message: string;
};

export const Error: FunctionComponent<Props> = ({ isVisible, message }) => {
	if (!isVisible) return null;

	return (
		<$Error>
			{message}
		</$Error>
	);
};
