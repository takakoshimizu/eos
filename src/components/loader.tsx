import React, { FunctionComponent } from 'react';
import styled, { keyframes } from 'styled-components';
import { color } from '../services/theme';

/* Thanks, https://loading.io/css/ as always */
const anim = keyframes`
	0% {
			top: 28px;
			left: 28px;
			width: 0;
			height: 0;
			opacity: 1;
		}
		100% {
			top: -1px;
			left: -1px;
			width: 58px;
			height: 58px;
			opacity: 0;
		}
`;

const $Loader = styled.div`
	display: inline-block;
	position: relative;
	width: 64px;
	height: 64px;

	div {
		position: absolute;
		border: 4px solid ${color('text')};
		opacity: 1;
		border-radius: 50%;
		animation: ${anim} 1s cubic-bezier(0, 0.2, 0.8, 1) infinite;
	}

	div:nth-child(2) {
		animation-delay: -0.5s;
	}
`;

const $LoaderHolder = styled.div`
	padding: 2rem;
	width: 64px;

	margin-left: auto;
	margin-right: auto;
`;

type Props = {
	isVisible: boolean;
};

export const Loader: FunctionComponent<Props> = ({ isVisible }) => {
	if (!isVisible) return null;

	return (
		<$LoaderHolder>
			<$Loader>
				<div />
				<div />
			</$Loader>
		</$LoaderHolder>
	);
};
