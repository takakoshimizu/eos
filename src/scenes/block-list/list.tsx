import React from 'react';
import styled from 'styled-components';
import { useSelector } from 'react-redux';

import { getBlocks } from '../../services/blocks/selectors';
import { Block } from '../../components/block';

const $List = styled.ul`
	list-style: none;
	padding: 0;
`;

export const List = () => {
	const blocks = useSelector(getBlocks);

	return (
		<$List>
			{blocks.map(b => <Block key={b.id} block={b} />)}
		</$List>
	);
};
