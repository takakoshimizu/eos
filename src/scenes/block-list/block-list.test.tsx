import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { I18nProvider } from '@lingui/react';

import { createStore } from 'redux';
import { reducers } from '../../store';
import { lightTheme } from '../../services/theme';
import en from '../../lang/en/messages';

import { BlockList } from '.';
import { List } from './list';
import { Loader } from '../../components/loader';
import { Error } from '../../components/error';
import { ThemeProvider } from 'styled-components';

const wrap = (initialState: any, story: any) => {
	const store = createStore(reducers, {
		blocks: [],
		fetchStatus: 'idle'
	});

	return (
		<Provider store={store}>
			<I18nProvider catalogs={{ en }} language="en">
				<ThemeProvider theme={lightTheme}>
					{story}
				</ThemeProvider>
			</I18nProvider>
		</Provider>
	)
};

describe('Block list', () => {
	it('should render', () => {
		const result = mount(
			wrap({ blocks: [], fetchStatus: 'idle' }, <BlockList />)
		);

		expect(result).toMatchSnapshot();
	});

	it('should render the loader when fetching', () => {
		const wrapper = mount(
			wrap({ blocks: [], fetchStatus: 'fetching' }, <BlockList />)
		);

		const result = wrapper.find(Loader);
		expect(result).toHaveLength(1);
	});

	it('should render the error message upon error', () => {
		const wrapper = mount(
			wrap({ blocks: [], fetchStatus: 'error' }, <BlockList />)
		);

		const result = wrapper.find(Error);
		expect(result).toHaveLength(1);
	});

	it('should render the list at all times', () => {
		const wrapper = mount(
			wrap({ blocks: [], fetchStatus: 'idle' }, <BlockList />)
		);

		const result = wrapper.find(List);
		expect(result).toHaveLength(1);
	});
});
