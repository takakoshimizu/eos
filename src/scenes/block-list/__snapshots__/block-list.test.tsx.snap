// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block list should render 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <I18nProvider
    catalogs={
      Object {
        "en": Object {
          "languageData": Object {
            "plurals": [Function],
          },
          "messages": Object {
            "EOSIO Blockchain Viewer": "EOSIO Blockchain Viewer",
          },
        },
      }
    }
    defaultRender={null}
    language="en"
  >
    <ThemeProvider
      theme={
        Object {
          "colors": Object {
            "accent": "#3277a8",
            "buttonText": "#ffffff",
            "card": "#f2f2f2",
            "error": "#a32121",
            "neutral": "#ffffff",
            "text": "#1a242d",
          },
          "font": Object {
            "primary": "Helvetica, sans-serif",
          },
          "shadows": Object {
            "z1": "0px 1px 3px 0px rgba(0,0,0,0.1), 0 1px 2px 0 rgba(0,0,0,0.2)",
            "z2": "0 3px 6px 0 rgba(0,0,0,0.23), 0 3px 6px 0 rgba(0,0,0,0.16)",
          },
        }
      }
    >
      <BlockList>
        <main>
          <h1>
            <WithI18n
              id="EOSIO Blockchain Viewer"
            >
              <I18n
                update={true}
                withHash={true}
              >
                <Trans
                  i18n={
                    I18n {
                      "_activeLanguageData": Object {
                        "plurals": [Function],
                      },
                      "_activeMessages": Object {
                        "EOSIO Blockchain Viewer": "EOSIO Blockchain Viewer",
                      },
                      "_catalogs": Object {
                        "en": Object {
                          "languageData": Object {
                            "plurals": [Function],
                          },
                          "messages": Object {
                            "EOSIO Blockchain Viewer": "EOSIO Blockchain Viewer",
                          },
                        },
                      },
                      "_dev": Object {
                        "compile": [Function],
                        "loadLanguageData": [Function],
                      },
                      "_language": "en",
                      "_locales": undefined,
                      "_missing": undefined,
                      "plural": [Function],
                      "select": [Function],
                      "selectOrdinal": [Function],
                      "t": [Function],
                    }
                  }
                  i18nHash={null}
                  id="EOSIO Blockchain Viewer"
                >
                  <Render
                    value="EOSIO Blockchain Viewer"
                  >
                    EOSIO Blockchain Viewer
                  </Render>
                </Trans>
              </I18n>
            </WithI18n>
          </h1>
          <styled.button
            onClick={[Function]}
          >
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bdVaJa",
                    "isStatic": false,
                    "lastClassName": "gDgQlX",
                    "rules": Array [
                      "
	padding: .5em 1em;
	border-radius: 6px;
	border: 0;

	box-shadow: ",
                      [Function],
                      ";

	background-color: ",
                      [Function],
                      ";
	color: ",
                      [Function],
                      ";

	transition: 
		background-color .3s ease-in-out,
		box-shadow .3s ease-in-out,
		transform .2s linear;

	&:hover {
		box-shadow: ",
                      [Function],
                      ";
		background-color: ",
                      [Function],
                      ";
		transform: translateY(-2px);
	}
",
                    ],
                  },
                  "displayName": "styled.button",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bdVaJa",
                  "target": "button",
                  "toString": [Function],
                  "usesTheme": true,
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
              onClick={[Function]}
            >
              <button
                className="sc-bdVaJa gDgQlX"
                onClick={[Function]}
              >
                <WithI18n
                  id="Get 10 most recent blocks"
                >
                  <I18n
                    update={true}
                    withHash={true}
                  >
                    <Trans
                      i18n={
                        I18n {
                          "_activeLanguageData": Object {
                            "plurals": [Function],
                          },
                          "_activeMessages": Object {
                            "EOSIO Blockchain Viewer": "EOSIO Blockchain Viewer",
                          },
                          "_catalogs": Object {
                            "en": Object {
                              "languageData": Object {
                                "plurals": [Function],
                              },
                              "messages": Object {
                                "EOSIO Blockchain Viewer": "EOSIO Blockchain Viewer",
                              },
                            },
                          },
                          "_dev": Object {
                            "compile": [Function],
                            "loadLanguageData": [Function],
                          },
                          "_language": "en",
                          "_locales": undefined,
                          "_missing": undefined,
                          "plural": [Function],
                          "select": [Function],
                          "selectOrdinal": [Function],
                          "t": [Function],
                        }
                      }
                      i18nHash={null}
                      id="Get 10 most recent blocks"
                    >
                      <Render
                        value="Get 10 most recent blocks"
                      >
                        Get 10 most recent blocks
                      </Render>
                    </Trans>
                  </I18n>
                </WithI18n>
              </button>
            </StyledComponent>
          </styled.button>
          <Loader
            isVisible={false}
          />
          <I18n
            update={true}
            withHash={true}
          >
            <Error
              isVisible={false}
              message="There was an error retrieving new blocks to display."
            />
          </I18n>
          <List>
            <styled.ul>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-dnqmqq",
                      "isStatic": false,
                      "lastClassName": "iruCdV",
                      "rules": Array [
                        "
	list-style: none;
	padding: 0;
",
                      ],
                    },
                    "displayName": "styled.ul",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-dnqmqq",
                    "target": "ul",
                    "toString": [Function],
                    "usesTheme": false,
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <ul
                  className="sc-dnqmqq iruCdV"
                />
              </StyledComponent>
            </styled.ul>
          </List>
        </main>
      </BlockList>
    </ThemeProvider>
  </I18nProvider>
</Provider>
`;
